lib/maths/CSampling.cc

```diff
@@ -191,45 +191,44 @@ void doCategoricalSampleWithoutReplacement(RNG& rng,
         return;
     }
 
-    std::size_t p = probabilities.size();
-    if (n >= p) {
-        result.assign(boost::counting_iterator<std::size_t>(0),
-                      boost::counting_iterator<std::size_t>(p));
+    std::size_t m{probabilities.size()};
+    result.assign(boost::counting_iterator<std::size_t>(0),
+                  boost::counting_iterator<std::size_t>(m));
+
+    if (n >= m) {
+        return;
     }
 
     // Construct the transform function.
-    for (std::size_t i = 1u; i < p; ++i) {
+    for (std::size_t i = 1; i < probabilities.size(); ++i) {
         probabilities[i] += probabilities[i - 1];
     }
 
-    result.reserve(n);
-    TSizeVec indices(boost::counting_iterator<std::size_t>(0),
-                     boost::counting_iterator<std::size_t>(p));
-    TSizeVec s(1);
-
-    for (std::size_t i = 0u; i < n; ++i, --p) {
-        if (probabilities[p - 1] <= 0.0) {
-            doUniformSample(rng, std::size_t(0), indices.size(), 1, s);
-            result.push_back(indices[s[0]]);
+    for (std::size_t i = 0; i < n; ++i, --m) {
+        std::size_t s{0};
+        double x{probabilities[m - 1]};
+        if (x <= 0.0) {
+            s = doUniformSample(rng, std::size_t{0}, m);
         } else {
-            boost::random::uniform_real_distribution<> uniform(0.0, probabilities[p - 1]);
-            double uniform0X = uniform(rng);
-            s[0] = std::min(static_cast<std::size_t>(
-                                std::lower_bound(probabilities.begin(),
-                                                 probabilities.end(), uniform0X) -
-                                probabilities.begin()),
-                            probabilities.size() - 1);
-
-            result.push_back(indices[s[0]]);
-
-            double ps = probabilities[s[0]] - (s[0] == 0 ? 0.0 : probabilities[s[0] - 1]);
-            for (std::size_t j = s[0] + 1; j < p; ++j) {
-                probabilities[j - 1] = probabilities[j] - ps;
-            }
-            probabilities.pop_back();
+            boost::random::uniform_real_distribution<> uniform{0.0, x};
+            double u0X{uniform(rng)};
+            s = std::min(static_cast<std::size_t>(
+                             std::lower_bound(probabilities.begin(),
+                                              probabilities.begin() + m, u0X) -
+                             probabilities.begin()),
+                         m - 1);
+        }
+
+        double ps{probabilities[s] - (s == 0 ? 0.0 : probabilities[s - 1])};
+        for (std::size_t j = s + 1; j < m; ++j) {
+            probabilities[j - 1] = probabilities[j] - ps;
+            std::swap(result[j - 1], result[j]);
         }
-        indices.erase(indices.begin() + s[0]);
     }
+
+    // The sampled values are at the end of the vector so copy into place.
+    std::copy(result.begin() + m, result.end(), result.begin());
+    result.resize(n);
```

Comment:
Would `result.erase(result.begin(), result.begin() + m);` be less efficient than these two lines?  It seems to me it would be doing basically the same thing but in one line.

Reply:
It guess it is implementation dependent, but any sane implementation would do what I do with the two lines...

Actually, the standard requires it to be "Linear on the number of elements erased (destructions) plus the number of elements after the last element deleted (moving).". This pretty much forces the hand of the implementation so should be as fast. Seems like it is worth tidying.

Reply:
See [d957489](https://github.com/elastic/ml-cpp/pull/1312/commits/d957489ee19d6eedc001373c3fac354aec037986).